New ego: Boots of lunging.

Implojin opened the pull request:
This commit adds SPARM_LUNGING and all the plumbing to go with it.
When equipped, this ego passively grants the player one "free" move action --
but only when moving directly towards a hostile monster.

Have fun!

ebering commented:
An interesting idea. Today is the 0.25 feature freeze, so this won't get in for 0.25 stable, but I'm very interested in trying it out in 0.26. Feel free to keep tweaking/rebasing/testing &c. during the feature-freeze-and-tournament period. If I have some time during then I'll try to give a more thorough review and test.

semitonal commented:
This currently has some broken interactactions with on-move effects such as barbs - you get your free square of lunge movement without taking any barb damage and then get a chance to cancel the rest of the move. Similarly the lunge portion of the movement doesn't cancel Ozocubu's Armour - I'd be a little worried about this requiring large amounts of code duplication.

In general I'm not a huge fan of the toggle interface - I see that it uses some similar implementation to permaflight, which I'd also really like to change to just be always-on and have no toggle (by removing the rare edge cases where you don't want to be flying). It seems maybe fine for these to be always-on too, and have the lunging be something you have to deal with in cases where it's a downside? Since they can't practically be swapped with enemies around. Also one minor detail is that either the ego or WJC's passive lunging would need a new name.

Implojin commented:
Thanks Kate, I had overlooked those effects with the initial commit.

I have mixed feelings re: removing the toggle interface for lunging:

From the perspective of "Yes, the intent of this ego is to hopefully provide some different movement pattern incentives", it's totally fine to make it always-on, and doing it that way would probably even make for better gameplay as it would eliminate any incentive to sit and consider the ability menu instead of just moving diagonally or whatever.

From the perspective of being approachable to newer players, though, I'm less convinced: The basic design of this ego encourages movement patterns that come with a lot of risk, and being able to situationally disable the effect might allow people to be more comfortable with actually equipping it instead of leaving it on the floor.

As I'm writing this, I think I agree with you, removing the toggle seems a bit more convincing. Anyway, removing that bit is pretty easy, I had it working that way initially before adding the toggle. I'll throw together another commit later to do that when I have the time.

Re: Bikeshedding the name: My initial draft of these was calling them 'dash boots', but I felt that implied too much freedom of movement. I also considered 'charging', but that sounded too much like an evocations thing. These work pretty well with stealth characters: Something like 'rogue's boots' or 'traveller's boots' might be apropos and okay? (If overlapping with boots of the assassin.) I'm open to suggestions, if you have any?

Implojin commented:
Okay, I've pushed a commit that removed the evoke toggle, hopefully I didn't miss any of that code.

If anyone has any further thoughts, I'd like to hear them?

PleasingFungus commented:
This is an extremely cool idea! Another +1 for trying this in 0.26.

ebering commented:
Well, the tournament has concluded and I'm back from vacation. There's a round of bug fixing and the 0.25.1 point release to do, but I'll be getting to this soon. If you can resolve the merge conflicts that would be nice but not necessary.

I think that if this ego is successful boots of running can run right out of the game, but I can handle that at a separate time.

ebering commented:
@Implojin I'm planning on giving this a review+test with an eye towards merging now that I've finished charms. Do you have time to rebase it on master and fix the merge conflicts? It would speed me up significantly.

Also @PleasingFungus will want to weigh in on the all important question: can a Palentonga in a lunging barding lunge roll charge?

PleasingFungus commented:
As in, does the entire roll become free? That seems like a bit of a can of worms, especially if slow melee weapons are involved... I'd sort of rather hold off on that, just as I've held off on WJC/charge and zerk/charge.

Implojin commented:
I'll take a look at the merge conflicts: If it's relatively minor enum stuff I might be able to resolve that tonight, otherwise I'm going to be unavailable until 2020/07/19.

I imagine the Palentonga stuff is likely to be thornier: If any of that requires resolution from my end, it will have to wait until Monday.

If possible, I'd like to request that you review that stuff in movement.cc in as explicit, nitpicky detail as possible before merging this: I recall not being super comfortable with the way that was written, and more eyes on it would be good before it gets into trunk.

Past that, there are a few minor changes to this PR that I still have in mind:

1) Disable lunging on naga bardings. This branch initially enabled Lunging for Centaur and Naga bardings, but after some playtesting and further thought, I'm leaning towards disabling this for Nagas. The low-cost repositioning kind of runs against the point of the Na species conduct.
2) Upload a better WIP tile. I got out the tablet and created a few different tiles for this, and while probably none of them are as good as whatever CanOfWorms might come up with, I have one or two that might work until someone creates something better.
3) Renaming the ego. I took a minute to [brainstorm some names](https://gist.github.com/Implojin/d746ddf9e306c388dd13e4011b98764a#file-alternate-lunging-names-txt) the other day, at present, I'm leaning towards "Rocinante". A close second might be "Carnage".

Implojin commented:
Okay, I've rebased this to master and cleaned up the merge conflicts. 

(I'm inexperienced with git, so I hope I haven't broken anything!)

There's not much left that needs doing here, I think: 

Palentonga compatibility probably needs to be fixed up (I haven't looked at any of the Palentonga movement code, but I doubt it's working properly with SPARM_LUNGING), and maybe some more bikeshedding of the ego name.

I've disabled SPARM_LUNGING from generating on Bardings for now, so you all can take your time deciding what the proper behavior there should be.

ebering commented:
Removing this ego from bardings solves the Palentonga problem as well by making it a non-issue.

FWIW Rolling Charge uses a separate code path from movement, so the current interaction would be "none".

Implojin commented on crawl-ref/source/movement.cc:
```diff
@@ -454,6 +502,251 @@ bool prompt_dangerous_portal(dungeon_feature_type ftype)
+        }
+    }
+    if (!valid_target)
+        return spret::fail;
```
Sure, no problem. Writing it that way does make the intended logic less ambiguous.

Implojin commented on crawl-ref/source/movement.cc:
```diff
@@ -454,6 +502,251 @@ bool prompt_dangerous_portal(dungeon_feature_type ftype)
+    if (feat_is_lava(grd(beam.target)))
+        return spret::fail;
+    else if (grd(beam.target) == DNGN_DEEP_WATER
+             || grd(beam.target) == DNGN_TOXIC_BOG)
```
At the time I wrote this, I had a concern that lunging over deep water / lava could potentially lead to unavoidable deaths due to flight expiry, as any valid lunge over these features would lead to the player being two tiles away from safe terrain.

4e7fcc has since removed the old instadeath codepath here, and this concern is now obviated, so, okay, sounds good.

ebering replied:
Welcome to the future, we've got fun and games

Implojin commented on crawl-ref/source/movement.cc:
```diff
@@ -454,6 +502,251 @@ bool prompt_dangerous_portal(dungeon_feature_type ftype)
+        // Don't lunge at invis mons, but allow the tracer to keep going.
+        else if (mon && !you.can_see(*mon))
+            continue;
+        // Don't lunge if the closest mons is non-hostile or a plant.
```
Ugh, this looks like it's going to add a lot of complexity.

ebering replied:
If you can make the changes other than Fedhas stuff I can take a crack at it.

Implojin replied:
Turns out it wasn't that bad, I've added Fedhas passthrough in 0752f75.

ebering replied:
Sweet!

Implojin commented on crawl-ref/source/player-equip.cc:
```diff
@@ -908,6 +908,10 @@ static void _equip_armour_effect(item_def& arm, bool unmeld,
             break;
+
+        case SPARM_LUNGING:
+            mpr("You feel ready to lunge towards enemies.");
```
Going to wait on the naming bikeshed to be resolved before fixing up any remaining messaging issues

Implojin commented on crawl-ref/source/item-name.cc:
```diff
@@ -567,6 +567,7 @@ const char* armour_ego_name(const item_def& item, bool terse)
 #endif
         case SPARM_HARM:              return "harm";
         case SPARM_SHADOWS:           return "shadows";
+        case SPARM_LUNGING:           return "carnage";
```
If you have any preferences I'm all ears. Many of the alternatives I considered don't succinctly convey both "motion" and "need for a target" inside of one verb that can also serve double duty as on-equip messaging. In my head, this is still kind of "lunging" at present.

Maybe it's not such a big deal to reuse this name from WJC?

Implojin replied:
For now, I've reverted the carnage naming. I honestly don't care what name ends up being used for this; I'll pop into ##crawl-dev and we can figure something out.

ebering replied:
Bikeshedding is eternal, I'll resolve this conversation and let that happen in the top level thread.

Implojin commented on crawl-ref/source/movement.cc:
```diff
@@ -454,6 +502,251 @@ bool prompt_dangerous_portal(dungeon_feature_type ftype)
+    {
+        // Don't lunge if our tracer path is broken by deep water, lava,
+        // teleport traps, etc., before it reaches a monster.
+        if (!feat_is_traversable(grd(p))
```
> and the logic used for `targ_pass` later in `move_player_action`.

Please clarify exactly what logic you want to add here: Are you talking about everything inside of the `if (!attacking && targ_pass && moving && !beholder && !fmonger)` braces?

ebering replied:
I was talking primarily about `targ_pass` at a cursory glance at your changes it looks correct.

Implojin replied:
I've replaced the logic in the tracer loop with `you.can_pass_through` and `you.see_cell_no_trans` calls; as far as the rest of the targ_pass logic goes that's applied further down in move_player_action, it looks to me like I'm already calling everything in `_lunge_forward` that's necessary for this? (I don't really think that enabling things like 2x confuse-stumbling or 2x constriction breaks per turn would be a good idea, here.)

ebering commented:
At a cursory glance (I know I said I was going to my office but it just started raining so gotta wait on biking) this looks good wrt. the things we discussed. In reviewing this and fixing up some charge code I adapted some of your factorization for water hold &c. and that's already been pulled out, which is why there's now a merge conflict. I'll handle fixing that up since I have notes from those changes handy still and there are some order consistency things.

Calling anyone from the sidelines who has a bikeshed opinion on the name for this ego. Here's some brainstorms to get you started: lunging, carnage, rampage, flèche.

RojjaCebolla commented:
Boots of Scooting, of Hitting First, of Surprise, of Opportunity, of Recklessness

Sillier:
of [5 Feet](https://www.dandwiki.com/wiki/SRD:5-Foot_Step), of Zooping Forwards, of Flutterfeet, of a Boost, of Hopping, of Smashing Red Devils, of Pitterpatter. Boots of Scoots

Implojin commented:
If we're doing silly names, I really liked boots of wizzardry

PleasingFungus commented:
Seven-league boots are almost, but not quite, apropos. No other unit of distance has quite the same ring, sadly.

Battle-seeking boots? Boots of battle? Warrior's boots, boots of soldiering?

Recklessness is fun, but might discourage new players from trying these. Rampage/rampaging might work.

Wizzardry is hilarious but feels more appropriate for running; these don't make you faster in general.

Implojin commented:
The beauty of wizzardry is that these will always be taking you towards more danger

I do kind of like seven-league boots

RojjaCebolla commented:
...of Courage, of the Bold, of Chutzpah and Moxie, of Swift Shanking, of Derring-do, of Daring, of the Jumping Spider, of Fleet Foxes...

ebering commented:
We're looking for ego names, not unrand names. One word is best, two tops, phrases are right out.

I'm sad flèche didn't get the response from the bikeshed gang I was hoping for.

Rampage and seven-league are good. The weapon ego description code currently supports adjective form egos, but this isn't generalized to armour ego (agh), so from an ease of completing point of view "boots of rampage" is favored there.

PleasingFungus commented:
(We do accept submissions for new randart names :) )

ebering commented:
After some play testing I think rampage fits well. A more elegant name would lead players to expect it to work with Wu; I'm going to add a note to Wu's extra screen that only normal movement is compatible with martial attacks (and remove dig lunging rip twelwe).

@Implojin brought up some concerns about rarity, especially versus running. Earlier in the conversation I mentioned these could replace the running ego (which is very strong bordering on no brainer territory). I won't do that for now but it's something to consider.

I have a local branch that cleans up the merge conflicts, squashes the history, and re-names them to rampaging (both in the code and in the player facing things, for the sake of some future developer trying to find something out about wu lunges).

It should land in trunk soon! Thanks for the work on this and the interesting idea.

RojjaCebolla commented:
Definitely doesn't need to be as rare as running, that ego is glorious

On Fri, Jul 24, 2020, 11:18 PM Edgar A. Bering IV <notifications@github.com>
wrote:

> After some play testing I think rampage fits well. A more elegant name
> would lead players to expect it to work with Wu; I'm going to add a note to
> Wu's extra screen that only normal movement is compatible with martial
> attacks (and remove dig lunging rip twelwe).
>
> @Implojin <https://github.com/Implojin> brought up some concerns about
> rarity, especially versus running. Earlier in the conversation I mentioned
> these could replace the running ego (which is very strong bordering on no
> brainer territory). I won't do that for now but it's something to consider.
>
> I have a local branch that cleans up the merge conflicts, squashes the
> history, and re-names them to rampaging (both in the code and in the player
> facing things, for the sake of some future developer trying to find
> something out about wu lunges).
>
> It should land in trunk soon! Thanks for the work on this and the
> interesting idea.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/crawl/crawl/pull/1409#issuecomment-663800646>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AO2KLERWACLCGXDMF66PKXTR5JFH3ANCNFSM4NCV33HA>
> .
>

ebering commented:
All merged up! Thanks again for the patch!
